cpp/src/arrow/array/validate.cc

```diff
@@ -509,22 +510,23 @@ struct ValidateArrayImpl {
     }
 
     if (full_validation) {
-      if (data.null_count != kUnknownNullCount) {
```

Comment:
I think we should still compute the null count only if needed by the checks, so e.g.:
```c++
      if (data.null_count != kUnknownNullCount || !nullable) {
```